### ISSUE TITLE
Incorrectly generated runpath and framework search path (Carthage)

### ISSUE THREAD

**Issue by ferologics** [tuist/tuist#2159]
**Describe the bug**
Tuist generates incorrect `LD_RUNPATH_SEARCH_PATHS` and `FRAMEWORK_SEARCH_PATHS` for Carthage dependencies. Both resolve into incorrect path `$(SRCROOT)/Carthage/Build/ios` with suffix `ios` instead of `iOS`.

**To Reproduce**
1. Download [TuistCarthage.zip](https://github.com/tuist/tuist/files/5716159/TuistCarthage.zip)
2. In project root run `tuist generate` (optional)
2. Open the TuistCarthage project
3. In Xcode navigator click on the project
4. Click on `TuistCarthageUITests` target
5. Notice `LD_RUNPATH_SEARCH_PATHS` is incorrect
4. Click on `Dependencies` target
5. Notice `FRAMEWORK_SEARCH_PATHS` is incorrect

**Expected behavior**
Tuist would generate a correct path `$(SRCROOT)/Carthage/Build/iOS`

**Screenshots**
<details>
<summary>Screenshots of incorrectly generated framework and runpath search paths</summary>

![image](https://user-images.githubusercontent.com/5576161/102624153-ec6b4780-4143-11eb-9a08-65bbbb0b2b25.png)
![image](https://user-images.githubusercontent.com/5576161/102624218-fee58100-4143-11eb-9710-2c67b4313b52.png)

</details>

**Desktop**

- OS: macOS `11.2 Beta (20D5029f)`
- Version `1.28.0`

**Additional context**

Probably an issue with the [`LinkGenerator`](https://github.com/tuist/tuist/blob/51ed4838fa3f57814ddbbdc7e60a46a730a670b1/Sources/TuistGenerator/Generator/LinkGenerator.swift#L292) and the usage of the `setup()` method:
```swift
try setup(setting: "LD_RUNPATH_SEARCH_PATHS",
                 paths: paths.map(LinkGeneratorPath.absolutePath),
                 pbxTarget: pbxTarget,
                 sourceRootPath: sourceRootPath)
```
where `SettingValue`:
```swift
let value = SettingValue
            .array(["$(inherited)"] + paths.map { $0.xcodeValue(sourceRootPath: sourceRootPath) }.uniqued().sorted())
```
where `xcodeValue`:
```swift
return "$(SRCROOT)/\(path.relative(to: sourceRootPath).pathString)"
```

**Comment by ferologics** [comment]
I figured it out. The `Platform.iOS.rawValue` evaluates to `ios` and not `iOS`. I used `platform.rawValue` when linking my Carthage frameworks as `.framework(path: Path("Carthage/Build/\(platform.rawValue)/\(name).framework"))`.

What's surprising is that everything was compiling just fine even though I linked all of my Carthage dependencies with `"Carthage/Build/ios/\(name).framework"`. Well, until I tried to run the tests. Then the incorrect `ios` path component for `LD_RUNPATH_SEARCH_PATHS` would fail the tests because the framework dependency was nowhere to be found. No wonder, when it was looking in the wrong place ☝️ 

I can see that `TuistCore` uses `caseValue` that correctly evaluates `Platform.iOS.caseValue` to `"iOS"`. I guess that internally Tuist is using it to evaluate the platform for linked Carthage dependencies? otherwise my project should not have compiled at all. I haven't found the exact place where it's doing the substitution. This is a pernicious implementation detail that is bound to cause more issues down the line.

I think that the `rawValue` should evaluate it's literal name or otherwise more fixtures will need to be put in place.

**Comment by pepibumur** [comment]
Thanks for the thorough analysis @ferologics 👏 

> What's surprising is that everything was compiling just fine even though I linked all of my Carthage dependencies with "Carthage/Build/ios/\(name).framework". Well, until I tried to run the tests. Then the incorrect ios path component for LD_RUNPATH_SEARCH_PATHS would fail the tests because the framework dependency was nowhere to be found. No wonder, when it was looking in the wrong place ☝️

That's a bit odd yeah. My first thought was the case-sensitivity of Apple File System (AFPS), however, that would have caused the compilation to fail as you pointed out. In the case of Carthage frameworks, we don't do any platform evaluation, we take the path that you provide, and then turn that into build settings, and the copy frameworks build phases. In your case that probably resolves into `Carthage/Build/ios/`. 

I think the build system ignores the case sensitivity, however, the copy build phase does not. Because of that, the framework is not copied and your app fails to launch. 

Does it work when you use the case value instead of the raw value? If it works I suggest [extending this model](https://github.com/tuist/tuist/blob/main/Sources/ProjectDescription/Platform.swift#L5) to make it conform `CaseIteratable` so that you can access `Platformn.iOS.caseValue` directly from the manifest.

**Comment by ferologics** [comment]
> I think the build system ignores the case sensitivity, however, the copy build phase does not. Because of that, the framework is not copied and your app fails to launch.

Maybe, thought I'm not sure how exactly runpath and framework search path build environment variables influence the above. The app compiled and ran on device w/o any issues. Only once I built the test target it failed to find the necessary frameworks.

> Does it work when you use the case value instead of the raw value?

Yes, locally I extended `Platform` with a `caseValue` computed property that returns the case sensitive literal value.

> If it works I suggest extending this model to make it conform `CaseIteratable` so that you can access Platformn.iOS.caseValue directly from the manifest.

I was thinking maybe we could extend the default `Platform` implementation with such property?

I'm not sure what is the motivation behind using lowercased `rawValue` but I'm sure you've got a good reason for it.

Alternative names for the `caseValue`:
- `literalValue`
- `caseSensitiveValue`
